Add `experimental` submodule

chaoming0625 opened the pull request:
In response to https://github.com/chaobrain/pinnx/issues/19, https://github.com/lululxvi/deepxde/discussions/1904, and https://github.com/lululxvi/deepxde/issues/1899, I am trying to merge [pinnx](https://github.com/chaobrain/pinnx) into `DeepXDE`. 

This pull request includes a submodule `deepxde.experimental`, which enables explicit variables and physical units in physics-informed neural networks. 



# Motivation


*experimental* module was engineered to address a pivotal challenge prevalent in current PINN libraries: the absence of explicit physical semantics. 

- Enhanced variable management with semantic clarity

Existing PINN libraries, such as *DeepXDE*, necessitate that users manually track the order and significance of variables. For instance, within these frameworks, `variables[0]` might denote amplitude, `variables[1]` for frequency, and so forth. This method lacks an intrinsic link between the sequence of variables and their physical meanings, thereby increasing complexity and the potential for errors. In stark contrast, *experimental* empowers users to assign clear and meaningful names to variables (e.g., `variables["amplitude"]` and `variables["frequency"]`). This approach obviates the need for manual management of variable order, enhancing both code readability and maintainability.

- Simplified gradient relationship management

Another significant limitation of existing PINN libraries is their reliance on users to manage intricate Jacobian and Hessian relationships between variables. This process is not only cumbersome but also prone to mistakes. *experimental* revolutionizes this aspect by employing a straightforward dictionary indexing mechanism to track intuitive gradient relationships. For example, users can effortlessly access the Hessian matrix element $\partial^2 y / \partial x^2$ via `hessian["y"]["x"]` and the Jacobian matrix element $\partial y / \partial t$ via `jacobian["y"]["t"]`. This simplification streamlines the workflow, allowing users to focus on modeling rather than matrix management.

- Integration of unit-aware automatic differentiation

Another prevalent deficiency in current PINN frameworks is the lack of support for physical units, which are crucial for maintaining dimensional consistency in physical equations. Take the Burgers' equation for example, the left-hand side $\frac{\partial u}{\partial t} + u\frac{\partial u}{\partial x}$ and the right-hand side $\nu\frac{\partial^2 u}{\partial x^2}$ must share the same physical units (e.g., meters per second squared). To ensure such consistency, *experimental* integrates `brainunit.autograd` module, enabling unit-aware automatic differentiation. This feature preserves unit information during the computation of first, second, or higher-order derivatives, ensuring that physical unit dimensions remain consistent throughout the differentiation process. Consequently, *experimental* effectively mitigates errors arising from unit mismatches, thereby enhancing the reliability of simulations.




# A quick example




```python

import brainstate as bst
import brainunit as u
from deepxde import experimental as deepxde

# geometry
geometry = deepxde.geometry.GeometryXTime(
    geometry=deepxde.geometry.Interval(-1, 1.),
    timedomain=deepxde.geometry.TimeDomain(0, 0.99)
).to_dict_point(x=u.meter, t=u.second)

uy = u.meter / u.second
v = 0.01 / u.math.pi * u.meter ** 2 / u.second

# boundary conditions
bc = deepxde.icbc.DirichletBC(lambda x: {'y': 0. * uy})
ic = deepxde.icbc.IC(lambda x: {'y': -u.math.sin(u.math.pi * x['x'] / u.meter) * uy})

# PDE equation
def pde(x, y):
    jacobian = approximator.jacobian(x)
    hessian = approximator.hessian(x)
    dy_x = jacobian['y']['x']
    dy_t = jacobian['y']['t']
    dy_xx = hessian['y']['x']['x']
    residual = dy_t + y['y'] * dy_x - v * dy_xx
    return residual

# neural network
approximator = deepxde.nn.Model(
    deepxde.nn.DictToArray(x=u.meter, t=u.second),
    deepxde.nn.FNN(
        [geometry.dim] + [20] * 3 + [1],
        "tanh",
        bst.init.KaimingUniform()
    ),
    deepxde.nn.ArrayToDict(y=uy)
)

# problem
problem = deepxde.problem.TimePDE(
    geometry,
    pde,
    [bc, ic],
    approximator,
    num_domain=2540,
    num_boundary=80,
    num_initial=160,
)

# training
trainer = deepxde.Trainer(problem)
trainer.compile(bst.optim.Adam(1e-3)).train(iterations=15000)
trainer.compile(bst.optim.LBFGS(1e-3)).train(2000, display_every=500)
trainer.saveplot(issave=True, isplot=True)
```

This new submodule support most of PINN examples in `DeepXDE`.

The documentation is hosted in `docs/experimental_docs` directory. The ample examples are included in `examples/experimental_examples`. 



# Installation

To use the `experimental` module, install it with the following command:

```bash 
pip install deepxde[experimental]
```



# Dependency

`experimental` module heavily depends on the following packages:

- [brainstate](https://github.com/chaobrain/brainstate): ``brainstate`` provides a state-based compilation system in JAX. It is very similar to pytorch's syntax. However, it can benefit from the JIT compilation of JAX and XLA. We recommend users to learn ``brainstate`` via its documentation: https://brainstate.readthedocs.io/quickstart/concepts-en.html
- [brainunit](https://github.com/chaobrain/brainunit): ``brainunit`` provides physical units and unit-aware mathematical systems tailored for Scientific AI within JAX. For the usage, please see its documentation: https://brainunit.readthedocs.io




# Low-precision training support

Changing the training precision is very easy in ``experimental``. Simply setting

```python
import brainstate as bst


bst.environ.set(precision='b16')
```

models will be trained using `bfloat16`. 


Here is an example:


```python

import brainstate as bst
import brainunit as u
import optax
from deepxde import experimental as deepxde

bst.environ.set(precision='b16')  # or, 32, 64, 16

geometry = deepxde.geometry.GeometryXTime(
    geometry=deepxde.geometry.Interval(-1, 1.),
    timedomain=deepxde.geometry.TimeDomain(0, 0.99)
).to_dict_point(x=u.meter, t=u.second)

uy = u.meter / u.second
bc = deepxde.icbc.DirichletBC(lambda x: {'y': 0. * uy})
ic = deepxde.icbc.IC(lambda x: {'y': -u.math.sin(u.math.pi * x['x'] / u.meter) * uy})

v = 0.01 / u.math.pi * u.meter ** 2 / u.second


def pde(x, y):
    jacobian = approximator.jacobian(x)
    hessian = approximator.hessian(x)
    dy_x = jacobian['y']['x']
    dy_t = jacobian['y']['t']
    dy_xx = hessian['y']['x']['x']
    residual = dy_t + y['y'] * dy_x - v * dy_xx
    return residual


approximator = deepxde.nn.Model(
    deepxde.nn.DictToArray(x=u.meter, t=u.second),
    deepxde.nn.FNN(
        [geometry.dim] + [20] * 3 + [1],
        "tanh",
    ),
    deepxde.nn.ArrayToDict(y=uy)
)

problem = deepxde.problem.TimePDE(
    geometry,
    pde,
    [bc, ic],
    approximator,
    num_domain=2540,
    num_boundary=80,
    num_initial=160,
)

trainer = deepxde.Trainer(problem, )
trainer.compile(bst.optim.OptaxOptimizer(optax.adamw(1e-3)))
trainer.train(iterations=10000)



```

chaoming0625 committed:
add `pinnx` submodule

chaoming0625 committed:
update

lululxvi commented:
Hi @chaoming0625 , thank you for adding the code. It is nice.

My first suggestion is that as there is some overlapped code between DeepXDE and PINNx, such as geometry, we should reduce the duplication as much as possible. You can import and reuse the code in DeepXDE.

chaoming0625 committed:
make changes

chaoming0625 commented:
Hi, @lululxvi , I have made some changes. However, `deepxde.pinnx.geometry` is necessary, since most of the points are generated using `jax.numpy` rather than `numpy`. These changes can enable the model compatible with `bfloat16`.

chaoming0625 commented:
In `deepxde.pinnx`, changing the traning precision is very easy. Setting

```python
import brainstate as bst


bst.environ.set(precision='b16')
```

Models will be trained using `bfloat16`. 


Here is an example:


```python

import brainstate as bst
import brainunit as u
import optax
import pinnx

bst.environ.set(precision='b16')  # or, 32, 64, 16

geometry = pinnx.geometry.GeometryXTime(
    geometry=pinnx.geometry.Interval(-1, 1.),
    timedomain=pinnx.geometry.TimeDomain(0, 0.99)
).to_dict_point(x=u.meter, t=u.second)

uy = u.meter / u.second
bc = pinnx.icbc.DirichletBC(lambda x: {'y': 0. * uy})
ic = pinnx.icbc.IC(lambda x: {'y': -u.math.sin(u.math.pi * x['x'] / u.meter) * uy})

v = 0.01 / u.math.pi * u.meter ** 2 / u.second


def pde(x, y):
    jacobian = approximator.jacobian(x)
    hessian = approximator.hessian(x)
    dy_x = jacobian['y']['x']
    dy_t = jacobian['y']['t']
    dy_xx = hessian['y']['x']['x']
    residual = dy_t + y['y'] * dy_x - v * dy_xx
    return residual


approximator = pinnx.nn.Model(
    pinnx.nn.DictToArray(x=u.meter, t=u.second),
    pinnx.nn.FNN(
        [geometry.dim] + [20] * 3 + [1],
        "tanh",
    ),
    pinnx.nn.ArrayToDict(y=uy)
)

problem = pinnx.problem.TimePDE(
    geometry,
    pde,
    [bc, ic],
    approximator,
    num_domain=2540,
    num_boundary=80,
    num_initial=160,
)

trainer = pinnx.Trainer(problem, )
_, compile_time = trainer.compile(
    bst.optim.OptaxOptimizer(optax.adamw(1e-3)),
    measture_train_step_compile_time=True
)
trainer.train(iterations=10000)



```

chaoming0625 commented:
Therefore, hosting `geometry` submodule is essential for `deepxde.pinnx` module.

chaoming0625 commented on deepxde/geometry/geometry.py:
```diff
@@ -4,8 +4,184 @@
 
 
-class Geometry(abc.ABC):
+class AbstractGeometry(abc.ABC):
```
Yes., it can. Sorry for this design.

chaoming0625 committed:
change the geometry hierarchy

chaoming0625 commented:
Hi, @lululxvi , in the next, my colleague will continue to promote the integration of 'pinnx' into 'deepxde', and he will focus more on solving every problem in the integration.

lululxvi commented:
> Hi, @lululxvi , in the next, my colleague will continue to promote the integration of 'pinnx' into 'deepxde', and he will focus more on solving every problem in the integration.

No problem.

lululxvi commented on deepxde/geometry/geometry.py:
```diff
@@ -6,6 +6,7 @@
 
 class Geometry(abc.ABC):
     def __init__(self, dim, bbox, diam):
+        super().__init__(dim)
```
This line is not needed.

chaoming0625 replied:
Fixed.

chaoming0625 committed:
fix bugs, check most models can successfully run

lululxvi commented on deepxde/pinnx/__init__.py:
```diff
@@ -0,0 +1,39 @@
+# Copyright 2024 BDP Ecosystem Limited. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
```
DeepXDE uses LGPL-2.1 license. Can we also use LGPL-2.1 license in these files?

chaoming0625 replied:
Sorry for this line. It should be changed to `DeepXDE`'s license.

lululxvi replied:
Then we just remove these comments

Routhleck replied:
Hi @lululxvi, I have just changed to LGPL-2.1 license in these files.

lululxvi replied:
How about we simply remove the license info in source code? As DeepXDE has the license file at https://github.com/lululxvi/deepxde/blob/master/LICENSE

Routhleck replied:
Sure, I have removed the liccense info in source code.

Routhleck committed:
Change to LGPL-2.1 license

Routhleck committed:
Updated copyright information to DeepXDE Limited

Routhleck committed:
Update installation.rst

Routhleck committed:
Remove license info in source code

lululxvi commented on deepxde/pinnx/fnspace.py:
```diff
@@ -0,0 +1,352 @@
+# Rewrite of the original file in DeepXDE: https://github.com/lululxvi/deepxde
```
Can we just use the file in DeepXDE?

chaoming0625 replied:
Yes, it is right. Sorry for the delay. I will fix it.

chaoming0625 replied:
I have removed `pinnx/fnspaces.py`, reusing functions in `deepxde/data` module. I have also tried examples in `examples/pinnx_examples/examples-operator`. They all work.

chaoming0625 committed:
Add optional dependencies for pinnx

we can install pinnx module using the following command:

pip install deepxde[pinnx]

chaoming0625 committed:
reuse functional spaces in `deepxde.data`

chaoming0625 commented:
I have updated the PR introduction to clearly explain the motivation behind the `pinnx` module. This should help users to understand the core purpose of the module.

chaoming0625 commented:
I have also added an option to install `pinnx` as an optional dependency, allowing users to quickly install the necessary components.  Specifically, users who wish to use the `pinnx` module can install it with the command:

```bash
pip install deepxde[pinnx]
```

chaoming0625 committed:
update dependency

chaoming0625 commented:
I updated the PR introduction once again.

chaoming0625 committed:
remove `pinnx/utils/sampler`, `pinnx/utils/sampling`, and `pinnx/utils/transformers`, reusing `deepxde` functions

chaoming0625 committed:
reuse `deepxde.callbacks`

chaoming0625 committed:
reuse `deepxde.model.LossHistory`

chaoming0625 committed:
reuse most functions in `deepxde.utils.internal`

chaoming0625 commented:
I have maximized function reuse from `deepxde` to minimize API redundancy.

chaoming0625 committed:
reuse models in `deepxde.nn.deeponet_strategy`

chaoming0625 committed:
add detailed docstring for most apis

chaoming0625 committed:
fix bugs, most models can run successfully

chaoming0625 commented:
I think now it is ready to merge this PR. I have tried most models. I also propose to leverage a more agile development process, in which we could consider submitting the PR in smaller increments.

chaoming0625 committed:
fix

chaoming0625 committed:
fix

chaoming0625 commented:
Hi, @lululxvi , could you check whether there are issues that need to be fixed? Now I have time to update the code.

lululxvi commented:
Sounds great! As this PR has a lot of changes, I will gradually review each file.

lululxvi commented on deepxde/callbacks.py:
```diff
@@ -556,6 +556,7 @@ def on_train_end(self):
                 )
 
 
+
```
No need to add this line.

chaoming0625 replied:
Fixed.

lululxvi commented on deepxde/nn/paddle/deeponet.py:
```diff
@@ -4,7 +4,7 @@
 from .. import activations
 from .. import initializers
-from ..deeponet_strategy import (
+from deepxde.utils.deeponet_strategy import (
```
Why using this kind of import? This import style is very different from lines 3-6.

chaoming0625 replied:
I have changed the code to keep the same style. 

One thing needs to note is that I have moved ``deepxde.nn.deeponet_strategy.py`` into ``deepxde.utils.deeponet_strategy.py``. This is because I cannot import this file in ``deepxde.pinnx.nn.deeponet.py``. It always reports the error ``ModuleNotFound``.

lululxvi replied:
> This is because I cannot import this file in `deepxde.pinnx.nn.deeponet.py`. It always reports the error `ModuleNotFound`.

This is strange. It is better to have the `deeponet_strategy.py` file in `deepxde/nn` as this file is only for nn. Is there another way to fix this without moving it to utils?

chaoming0625 replied:
Fixed this issue by importing `deeponet_strategy` in `deepxde.nn` module.

chaoming0625 committed:
fix code

chaoming0625 committed:
import `deeponet_strategy` in `deepxde.nn` module

lululxvi commented on deepxde/nn/__init__.py:
```diff
@@ -37,4 +37,5 @@ def _load_backend(mod_name):
         setattr(thismod, api, obj)
 
 
+from . import deeponet_strategy
```
```
from . import deeponet_strategy
from ..backend import backend_name
```

chaoming0625 replied:
Fixed

lululxvi commented on README.md:
````diff
@@ -98,6 +98,13 @@ $ conda install -c conda-forge deepxde
+``` sh
+$ pip install deepxde[pinnx]
+```
+
````
Also here, only one black line.

lululxvi commented on pyproject.toml:
```diff
@@ -54,3 +54,15 @@ exclude = ["docker", "docs*", "examples*"]
 [tool.setuptools_scm]
 write_to = "deepxde/_version.py"
+
+
```
Only one blank line.

chaoming0625 replied:
Fixed

chaoming0625 committed:
import `deeponet_strategy` in `__init__.py`

chaoming0625 committed:
update README to include installation instructions for pinnx module

chaoming0625 commented on README.md:
````diff
@@ -98,6 +98,13 @@ $ conda install -c conda-forge deepxde
 $ git clone https://github.com/lululxvi/deepxde.git
 ```
 
+
````
Fixed

chaoming0625 committed:
remove unnecessary blank line in pyproject.toml

lululxvi commented:
Why this module is named as "pinnx"?

chaoming0625 commented:
This is a great question! Could you suggest a better name?

chaoming0625 commented:
I agree this is a good name. If this name is applied, how can we import this module?

```python
import deepxde.experimental as deepxde

```

lululxvi commented:
> ```python
> import deepxde.experimental as deepxde
> ```

This looks good to me.

chaoming0625 committed:
refactor: rename pinnx module to deepxde.experimental and update imports

chaoming0625 commented:
> > ```python
> > import deepxde.experimental as deepxde
> > ```
> 
> This looks good to me.

I have made all necessary updates by renaming `deepxde.pinnx` to `deepxde.experimental`. Also updated all corresponding scripts.

chaoming0625 committed:
refactor: rename pinnx examples and documentation files to improve organization

lululxvi commented:
> I have made all necessary updates by renaming `deepxde.pinnx` to `deepxde.experimental`. Also updated all corresponding scripts.

Sounds good. But I realize that you changed the code without reusing the DeepXDE code. It is still better to reuse existing code as many as possible.

chaoming0625 commented:
> > I have made all necessary updates by renaming `deepxde.pinnx` to `deepxde.experimental`. Also updated all corresponding scripts.
> 
> Sounds good. But I realize that you changed the code without reusing the DeepXDE code. It is still better to reuse existing code as many as possible.

I'm unclear on what you're referring to. Code reuse problems were resolved across many previous commits. This commit's sole purpose is to change the module name.

lululxvi commented on deepxde/experimental/_trainer.py:
```diff
@@ -0,0 +1,508 @@
+# Rewrite of the original file in DeepXDE: https://github.com/lululxvi/deepxde
```
Why this file name is `_trainer`? not `trainer`?

chaoming0625 replied:
The prefix `_*` means the internal module, which cannot be accessed by `deepxde.experimental.trainer.XXXX`. Instead, users should use `deepxde.experimental.XXXX`.

lululxvi commented on deepxde/experimental/_trainer.py:
```diff
@@ -0,0 +1,508 @@
+# Rewrite of the original file in DeepXDE: https://github.com/lululxvi/deepxde
```
Remove the comments
```
# Rewrite of the original file in DeepXDE: https://github.com/lululxvi/deepxde
# ==============================================================================
```

chaoming0625 committed:
refactor: remove redundant comments from multiple files

lululxvi commented:
Please format the code using black https://github.com/psf/black

chaoming0625 committed:
refactor: remove unnecessary blank lines and improve code formatting across multiple files

chaoming0625 commented:
> Please format the code using black https://github.com/psf/black

Done

chaoming0625 commented:
Hi, @lululxvi can this PR be merged now? Or are there any other issues that need to be solved?

chaoming0625 commented:
Sorry. It's really hard to make such a split. I think this PR has a very good fit and quality.

Routhleck commented:
Hi @lululxvi, I just spilt this pr into #2003 #2004 #2005. Hope this works for you.

echen5503 commented:
@lululxvi Recommend close this PR in favor of #2003, #2004, #2005

chaoming0625 commented:
> @lululxvi Recommend close this PR in favor of #2003, #2004, #2005

OK, we will make update.
